pkg/util/sketch/topk.go

```diff
@@ -0,0 +1,248 @@
+	sketch              *CountMinSketch
+	bf                  [][]bool
+	hashfuncs, rowlen   int
+	hll                 *hyperloglog.Sketch
+	expectedCardinality int
+}
+
+// get the correct sketch width based on the expected cardinality of the set
+// we might need to do something smarter here to round up to next order of magnitude if we're say more than 10%
+// over a given size that currently exists, or have some more intermediate sizes
+func getCMSWidth(l log.Logger, c int) int {
+	// default to something reasonable for low cardinality
+	width := 32
+	switch {
+	case c >= 1000001:
+		if l != nil {
+			level.Warn(l).Log("cardinality is greater than 1M but we don't currently have predefined sketch sizes for cardinalities that large")
+		}
+	case c >= 1000000:
+		width = 409600
+	case c >= 100000:
+		width = 65536
+	case c >= 10000:
+		width = 5120
+	case c >= 1000:
+		width = 640
+	case c >= 100:
+		width = 48
+	}
+	return width
+}
+
+func NewCMSTopkForCardinality(l log.Logger, k, c int) (*Topk, error) {
+	// TODO: fix this function and get width function based on new testing data
+	// a depth of > 4 didn't seem to make things siginificantly more accurate during testing
+	w := getCMSWidth(l, c)
+	d := 4
+
+	sk, err := NewCMSTopK(k, w, d)
+	if err != nil {
+		return &Topk{}, err
+	}
+	sk.expectedCardinality = c
+	return sk, nil
+}
+
+func NewCMSTopK(k, w, d int) (*Topk, error) {
+	s, err := NewCountMinSketch(w, d)
+	if err != nil {
+		return &Topk{}, nil
+	}
+	return &Topk{
+		max:    k,
+		heap:   &MinHeap{},
+		sketch: s,
+		hll:    hyperloglog.New16(),
+		bf:     makeBF(w, d),
+	}, nil
+}
+
+func (t *Topk) heapPush(event string, estimate, h1, h2 uint32) {
+	var pos uint32
+	for i := range t.bf {
+		pos = t.sketch.getPos(h1, h2, i)
+		t.bf[i][pos] = true
+	}
+	heap.Push(t.heap, &node{event: event, count: estimate})
+}
+
+func (t *Topk) heapUpdate(event string, estimate, h1, h2 uint32) {
+	var pos uint32
+	for i := range t.bf {
+		pos = t.sketch.getPos(h1, h2, i)
+		t.bf[i][pos] = true
+	}
+	(*t.heap)[0].event = event
+	(*t.heap)[0].count = estimate
+	heap.Fix(t.heap, 0)
+}
+
+// updates the BF to ensure that the removed event won't be mistakenly thought
+// to be in the heap, and updates the BF to ensure that we would get a truthy result for the added event
+func (t *Topk) updateBF(removed, added string) {
+	r1, r2 := hashn(removed)
+	a1, a2 := hashn(added)
+	var pos uint32
+	for i := range t.bf {
+		// removed event
+		pos = t.sketch.getPos(r1, r2, i)
+		t.bf[i][pos] = false
+		// added event
+		pos = t.sketch.getPos(a1, a2, i)
+		t.bf[i][pos] = true
+	}
+}
+
+func (t *Topk) Observe(event string) {
+	estimate, h1, h2 := t.sketch.ConservativeIncrement(event)
+	t.hll.InsertHash(uint64(h1))
+
+	// check if the event is already in the topk, if it is we should update it's count
+	if t.InTopk(h1, h2) {
+		return
+	}
+
+	if len(*t.heap) < t.max {
+		t.heapPush(event, estimate, h1, h2)
+		return
+	}
+
+	if estimate > t.heap.Peek().(*node).count {
```

Comment:
I'm not sure I understand this logic. Why are we looking at the minimum of the heap?

Reply:
I'll add a [link](http://www.eecs.harvard.edu/~michaelm/postscripts/tr-02-05.pdf) to a comment above the `Observe` function to explain the source of the optimization here, but the short(ish) version is this:

By default when we observe an event, if it's already in the current topk we would update it's value in the heap structure with the new count min sketch estimate and then rebalance the heap. This is potentially a lot of heap balancing operations that, at the end of the day, aren't really important. What information do we care about from the heap when we're actually still observing events and tracking the topk? The minimum value that's stored in the heap. If we observe an event and it's new count is greater than the minimum value in the heap, that event should go into the heap and the event with the minimum value should come out.

The optimization here is that we don't need to update the values for each item in the heap each time we observe an event that's in the heap, we only need to update them when we observe an event that's **not** in the heap, and it's new estimate is **greater** than the thing that's the current minimum value heap element. At that point, we update the values for each node in the heap and rebalance the heap, and then if the event we're observing has an estimate that is **still** greater than the minimum heap element count, we should put this event into the heap and remove the other one. This is the optimization that sped up the benchmark by ~3x.

Reply:
Nice. Maybe we could add that as a comment. I've seen algorithm implementations with lengthy comments. I think it's totally valid to put your explanation into the code base.

Reply:
I've added a comment to the topk structure as well as above the `Observe` function about the optimizations, let me know if you think they explain it well enough.